compiler/parser/parser.cpp

```diff
@@ -1069,34 +1090,63 @@ static DynoErrorHandler* dynoPrepareAndInstallErrorHandler(void) {
 // Only install one of these for the entire session.
 static DynoErrorHandler* gDynoErrorHandler = nullptr;
 
+static bool dynoRealizeError(const chpl::owned<chpl::ErrorBase>& err) {
+  bool hadErrors = false;
+  const chpl::ErrorBase* e = err.get();
+
+  if (e->kind() == chpl::ErrorBase::SYNTAX ||
+      e->kind() == chpl::ErrorBase::ERROR) {
+    // make a note if we found any errors
+    hadErrors = true;
+  }
+
+  // skip emitting warnings for '--no-warnings'
+  if (ignore_warnings && e->kind() == chpl::ErrorBase::WARNING) {
+    return hadErrors;
+  }
+
+  if (fDetailedErrors) {
+    chpl::Context::defaultReportError(gContext, e);
+    // Use production compiler's exit-on-error functionality for errors
+    // reported via new Dyno mechanism
+    setupDynoError(e->kind());
+  } else {
+    // Try to maintain compatibility with the old reporting mechanism
+    dynoDisplayError(gContext, e->toErrorMessage(gContext));
+  }
+  return hadErrors;
+}
+
 static bool dynoRealizeErrors(void) {
   INT_ASSERT(gDynoErrorHandler);
   bool hadErrors = false;
   for (auto& err : gDynoErrorHandler->errors()) {
-    const chpl::ErrorBase* e = err.get();
+    hadErrors |= dynoRealizeError(err);
+  }
+  gDynoErrorHandler->clear();
+  return hadErrors;
+}
 
-    if (e->kind() == chpl::ErrorBase::SYNTAX ||
-        e->kind() == chpl::ErrorBase::ERROR) {
-      // make a note if we found any errors
-      hadErrors = true;
-    }
+static bool dynoRealizeDeferredErrors(void) {
+  INT_ASSERT(gDynoErrorHandler);
+  bool hadErrors = false;
+  auto mainModulePath = ModuleSymbol::mainModule()->path();
 
-    // skip emitting warnings for '--no-warnings'
-    if (ignore_warnings && e->kind() == chpl::ErrorBase::WARNING) {
-      continue;
-    }
+  for (auto& err : gDynoErrorHandler->deferredErrors()) {
+    if (err->type() == chpl::ErrorType::ImplicitFileModule) {
+      auto errCast = (const chpl::ErrorImplicitFileModule*) err.get();
+      auto implicitMod = std::get<2>(errCast->info());
 
-    if (fDetailedErrors) {
-      chpl::Context::defaultReportError(gContext, e);
-      // Use production compiler's exit-on-error functionality for errors
-      // reported via new Dyno mechanism
-      setupDynoError(e->kind());
-    } else {
-      // Try to maintain compatibility with the old reporting mechanism
-      dynoDisplayError(gContext, e->toErrorMessage(gContext));
+      if (implicitMod->id().symbolPathWithoutRepeats(gContext) == mainModulePath) {
```

Comment:
What information is contained in the symbol path and main module path? Like, does it include an absolute path and then module/submodule or just the module info? Thinking about behavior when the user gives the mainModule as an absolute or relative path on the command line.

Reply:
The symbol path and module path are both filename-independent. The symbol path is a bit of a heuristic when it comes to Dyno (some "symbols" like variables don't get their own path, others like modules and functions do). However (I have verified) that modules can only be nested inside of other modules, and thus that the symbol path for a module will include the chain of modules from the root (e.g., `M.N.O`). The only reason it might differ is due to duplication (`module M {}; module M {}`), in which case one might have symbol path `M#0` . This is the reason I've added `symbolPathWithoutRepeats`. 

The `mainModulePath` as a command line argument, is usually specified as `--main-module A` (as opposed to `--main-module A.chpl`). You can confirm this by noting that it's usually compared against `ModuleSymbol::path`, which is computed using names only:

```C++
std::string ModuleSymbol::path() const {
  std::string retval;

  if (this == rootModule) {
    retval = name;

  } else {
    ModuleSymbol* parent = toModuleSymbol(defPoint->parentSymbol);

    if (parent == theProgram) {
      retval = name;
    } else {
      retval = (parent->path() + ".") + name;
    }
  }

  return retval;
}
```

Thus, we have two filename-independent (and thus "absolute"/"relative"-independent) variables here.

Reply:
Awesome, thank you!